### PR TITLE
fix(buttons): Fixes handling for disabled attribute and property

### LINKED ISSUE #1088: NgbRadio disabled does not work.
Adding the disabled attribute to an input is removed on page load. Success in tests however it appears that the browser does two passes, removing this attribute on the second.

http://plnkr.co/edit/9udhxszOhPKZAEWcccmi

## Versions:
Angular: 2.0.0, 2.1.1
NGbootstrap: 1.0.0-alpha.9, 1.0.0-alpha.14
Bootstrap: 4.0.0-alpha.5, 4.0.0-alpha.2

## Stacktrace:
x | x
---|---
DomRenderer.setElementAttribute                       | @ platform-browser.umd.js:1458
DomRenderer.setBindingDebugInfo                       | @ platform-browser.umd.js:1477
DebugDomRenderer.setBindingDebugInfo                  | @ core.umd.js:8850
setBindingDebugInfo                                   | @ core.umd.js:5433
Wrapper_NgbRadio.detectChangesInHostProps             | @ wrapper.ngfactory.js:61
_View_CoverOptionsComponent0.detectChangesInternal    | @ component.ngfactory.js:2006
AppView.detectChanges                                 | @ core.umd.js:9305
DebugAppView.detectChanges                            | @ core.umd.js:9410
AppView.detectViewChildrenChanges                     | @ core.umd.js:9331
_View_CustomiseComponent2.detectChangesInternal       | @ component.ngfactory.js:1058
AppView.detectChanges                                 | @ core.umd.js:9305
DebugAppView.detectChanges                            | @ core.umd.js:9410
AppView.detectContentChildrenChanges                  | @ core.umd.js:9323
_View_NgbAccordion3.detectChangesInternal             | @ component.ngfactory.js:255
AppView.detectChanges                                 | @ core.umd.js:9305
DebugAppView.detectChanges                            | @ core.umd.js:9410
AppView.detectContentChildrenChanges                  | @ core.umd.js:9323
_View_NgbAccordion1.detectChangesInternal             | @ component.ngfactory.js:157
AppView.detectChanges                                 | @ core.umd.js:9305
DebugAppView.detectChanges                            | @ core.umd.js:9410
AppView.detectContentChildrenChanges                  | @ core.umd.js:9323
_View_NgbAccordion0.detectChangesInternal             | @ component.ngfactory.js:58
AppView.detectChanges                                 | @ core.umd.js:9305
DebugAppView.detectChanges                            | @ core.umd.js:9410
AppView.detectViewChildrenChanges                     | @ core.umd.js:9331
_View_CustomiseComponent0.detectChangesInternal       | @ component.ngfactory.js:840
AppView.detectChanges                                 | @ core.umd.js:9305
DebugAppView.detectChanges                            | @ core.umd.js:9410
AppView.detectViewChildrenChanges                     | @ core.umd.js:9331
_View_CustomiseComponent_Host0.detectChangesInternal  | @ host.ngfactory.js:33
AppView.detectChanges                                 | @ core.umd.js:9305
DebugAppView.detectChanges                            | @ core.umd.js:9410
ViewRef_.detectChanges                                | @ core.umd.js:7398
RouterOutlet.activate                                 | @ router.umd.js:3458
ActivateRoutes.placeComponentIntoOutlet               | @ router.umd.js:2955
ActivateRoutes.activateRoutes                         | @ router.umd.js:2933
(anonymous function)                                  | @ router.umd.js:2902
ActivateRoutes.activateChildRoutes                    | @ router.umd.js:2901
ActivateRoutes.activate                               | @ router.umd.js:2896
(anonymous function)                                  | @ router.umd.js:2646
(anonymous function)                                  | @ Observable.js:108
SafeSubscriber.__tryOrSetError                        | @ Subscriber.js:232
SafeSubscriber.next                                   | @ Subscriber.js:174
Subscriber._next                                      | @ Subscriber.js:125
Subscriber.next                                       | @ Subscriber.js:89
MergeMapSubscriber.notifyNext                         | @ mergeMap.js:133
InnerSubscriber._next                                 | @ InnerSubscriber.js:23
Subscriber.next                                       | @ Subscriber.js:89
MapSubscriber._next                                   | @ map.js:83
Subscriber.next                                       | @ Subscriber.js:89
ReduceSubscriber._complete                            | @ reduce.js:105
Subscriber.complete                                   | @ Subscriber.js:114
MergeMapSubscriber._complete                          | @ mergeMap.js:125
Subscriber.complete                                   | @ Subscriber.js:114
ArrayObservable._subscribe                            | @ ArrayObservable.js:116
Observable.subscribe                                  | @ Observable.js:56
Observable._subscribe                                 | @ Observable.js:114
MergeMapOperator.call                                 | @ mergeMap.js:75
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
ReduceOperator.call                                   | @ reduce.js:63
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
subscribeToResult                                     | @ subscribeToResult.js:21
MergeMapSubscriber._innerSub                          | @ mergeMap.js:120
MergeMapSubscriber._tryNext                           | @ mergeMap.js:117
MergeMapSubscriber._next                              | @ mergeMap.js:100
Subscriber.next                                       | @ Subscriber.js:89
MergeMapSubscriber.notifyNext                         | @ mergeMap.js:133
InnerSubscriber._next                                 | @ InnerSubscriber.js:23
Subscriber.next                                       | @ Subscriber.js:89
EverySubscriber.notifyComplete                        | @ every.js:47
EverySubscriber._complete                             | @ every.js:64
Subscriber.complete                                   | @ Subscriber.js:114
MergeAllSubscriber._complete                          | @ mergeAll.js:94
Subscriber.complete                                   | @ Subscriber.js:114
Subscriber._complete                                  | @ Subscriber.js:132
Subscriber.complete                                   | @ Subscriber.js:114
ArrayObservable._subscribe                            | @ ArrayObservable.js:116
Observable.subscribe                                  | @ Observable.js:56
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MergeAllOperator.call                                 | @ mergeAll.js:63
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
EveryOperator.call                                    | @ every.js:27
Observable.subscribe                                  | @ Observable.js:53
subscribeToResult                                     | @ subscribeToResult.js:21
MergeMapSubscriber._innerSub                          | @ mergeMap.js:120
MergeMapSubscriber._tryNext                           | @ mergeMap.js:117
MergeMapSubscriber._next                              | @ mergeMap.js:100
Subscriber.next                                       | @ Subscriber.js:89
MapSubscriber._next                                   | @ map.js:83
Subscriber.next                                       | @ Subscriber.js:89
MapSubscriber._next                                   | @ map.js:83
Subscriber.next                                       | @ Subscriber.js:89
MapSubscriber._next                                   | @ map.js:83
Subscriber.next                                       | @ Subscriber.js:89
MergeMapSubscriber.notifyNext                         | @ mergeMap.js:133
InnerSubscriber._next                                 | @ InnerSubscriber.js:23
Subscriber.next                                       | @ Subscriber.js:89
subscribeToResult                                     | @ subscribeToResult.js:16
MergeMapSubscriber._innerSub                          | @ mergeMap.js:120
MergeMapSubscriber._tryNext                           | @ mergeMap.js:117
MergeMapSubscriber._next                              | @ mergeMap.js:100
Subscriber.next                                       | @ Subscriber.js:89
Subscriber._next                                      | @ Subscriber.js:125
Subscriber.next                                       | @ Subscriber.js:89
MapSubscriber._next                                   | @ map.js:83
Subscriber.next                                       | @ Subscriber.js:89
MapSubscriber._next                                   | @ map.js:83
Subscriber.next                                       | @ Subscriber.js:89
MapSubscriber._next                                   | @ map.js:83
Subscriber.next                                       | @ Subscriber.js:89
LastSubscriber._complete                              | @ last.js:109
Subscriber.complete                                   | @ Subscriber.js:114
MergeAllSubscriber._complete                          | @ mergeAll.js:94
Subscriber.complete                                   | @ Subscriber.js:114
ScalarObservable._subscribe                           | @ ScalarObservable.js:51
Observable.subscribe                                  | @ Observable.js:56
Observable._subscribe                                 | @ Observable.js:114
MergeAllOperator.call                                 | @ mergeAll.js:63
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
LastOperator.call                                     | @ last.js:38
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
CatchOperator.call                                    | @ catch.js:30
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MergeMapOperator.call                                 | @ mergeMap.js:75
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MapOperator.call                                      | @ map.js:54
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MergeMapOperator.call                                 | @ mergeMap.js:75
Observable.subscribe                                  | @ Observable.js:53
Observable._subscribe                                 | @ Observable.js:114
MergeMapOperator.call                                 | @ mergeMap.js:75
Observable.subscribe                                  | @ Observable.js:53
(anonymous function)                                  | @ Observable.js:87
ZoneAwarePromise                                      | @ zone.js:518
Observable.forEach                                    | @ Observable.js:86
(anonymous function)                                  | @ router.umd.js:2630
ZoneAwarePromise                                      | @ zone.js:518
Router.runNavigate                                    | @ router.umd.js:2595
(anonymous function)                                  | @ router.umd.js:2586
ZoneDelegate.invoke                                   | @ zone.js:232
onInvoke                                              | @ core.umd.js:6206
ZoneDelegate.invoke                                   | @ zone.js:231
Zone.run                                              | @ zone.js:114
(anonymous function)                                  | @ zone.js:502
ZoneDelegate.invokeTask                               | @ zone.js:265
onInvokeTask                                          | @ core.umd.js:6197
ZoneDelegate.invokeTask                               | @ zone.js:264
Zone.runTask                                          | @ zone.js:154
drainMicroTaskQueue                                   | @ zone.js:401

### LINKED ISSUE #1089: radio buttons cannot be re-enabled after disabled
### Bug description:
radio buttons can be disabled as expected. But it cannot be re-enabled. This does not happen without using ng-bootstrap.
### Link to minimally-working plunker that reproduces the issue:
http://plnkr.co/edit/TO4Y6HsxMvg1urYdBDX4?p=preview
In plnkr, textbox behave normally.

### Version of Angular, ng-bootstrap, and Bootstrap:

Angular: 2.2.1

ng-bootstrap: latest

Bootstrap: 4.0

### PR DESCRIPTION
**Pull request by alex321**
Fixes #1088
Fixes #1089

### CODE REVIEW

**Comment by alex321** [comment]
@pkozlowski-opensource Let me know what you think. It works, but I guess it could be done better. I had a look at Angular Material 2 for comparison and ideas.

**Comment by alex321** [comment]
Added test and implemented the following: 
- if `ngbRadioGroup` is disabled then all inputs are disabled
- if `ngbRadioGroup` is enabled then inputs are disabled independently

I hope that this should fix all relevant functionality.

File: src/buttons/radio.ts [hunk 1]
```diff
@@ -139,10 +160,10 @@ export class NgbRadio implements OnDestroy {
...
-  update(value, isDisabled) {
+  updateValue(value: any) {
     this._checked = (this.value === value && value !== null);
-    this._disabled = isDisabled;
-    this._label.active = this._checked;
-    this._label.disabled = this._disabled;
+    this._label.active = this.checked;
   }
+
+  updateDisableState() { this._crf.detectChanges(); }
```
Comment (pkozlowski-opensource) [inline src/buttons/radio.ts]: Wow, this sounds super-fishy! Why do we need to do this?

File: src/buttons/radio.ts [hunk 2]
```diff
@@ -117,11 +132,17 @@ export class NgbRadio implements OnDestroy {
 
   get checked() { return this._checked; }
 
-  get disabled() { return this._disabled; }
+  get disabled() {
+    let result = (this._group && this._group.disabled) || this._disabled;
+    if (this._label) {
+      this._label.disabled = result;
```
Comment (pkozlowski-opensource) [inline src/buttons/radio.ts]: I'm not super-fan of having side-effects in a getter (in other words - we shouldn't be doing this). What is the motivation / problem that we are trying to solve here?

**Comment by pkozlowski-opensource** [comment]
@alex321 I had a look at the PR and while tests look good I've got some reservations regarding the existing implementation (left comments in the PR). 

Also I would prefer that we separate fixing a bug (handling of the `disabled` state on a button level) from new features (adding `disabled` support on a group level). 

Given the 2 above comments I'm not keen on merging this PR as-is. Do you think you can give it another shoot with just a bug fix and avoiding impl patterns I've pointed out?

**Comment by pkozlowski-opensource** [comment]
> Also I would prefer that we separate fixing a bug (handling of the disabled state on a button level) from new features (adding disabled support on a group level).

Scratch this, we do already have support for the disabled state on the group level. Sorry for the noise.

File: src/buttons/radio.ts (hunk 2 above)
Comment (alex321) [inline src/buttons/radio.ts]: I didn't like that either. It is a bit similar to the previous comment. 
The question is how to keep the label in sync with the `input`. Again, I will have a look, but any suggestions appreciated.

**Comment by pkozlowski-opensource** [comment]
@alex321 in fact I've played with it a bit now and I think I've got an alternative approach - could you please have a look at #1141 and let me know what you think?

**Pull request closed without merging**